tests/catch/unit/memory/hipFree.cc

```diff
@@ -0,0 +1,522 @@
+
+// Freeing a invalid pointer with array
+TEST_CASE("Unit_hipFreeNegativeArray") {
+  hipArray_t arrayPtr{};
+
+  SECTION("InvalidPtr") {
+    arrayPtr = static_cast<hipArray*>(malloc(sizeof(char)));
+
+    SECTION("ArrayFree") { HIP_CHECK_ERROR(hipFreeArray(arrayPtr), hipErrorContextIsDestroyed); }
+    SECTION("ArrayDestroy") {
+      HIP_CHECK_ERROR(hipArrayDestroy(arrayPtr), hipErrorContextIsDestroyed);
+    }
+  }
+
+  SECTION("NullPtr") {
+    arrayPtr = nullptr;
+    SECTION("ArrayFree") { HIP_CHECK_ERROR(hipFreeArray(arrayPtr), hipErrorInvalidValue); }
+    SECTION("ArrayDestroy") { HIP_CHECK_ERROR(hipArrayDestroy(arrayPtr), hipErrorInvalidValue); }
+  }
+
+  free(arrayPtr);
+}
+
+#endif
+
+
+TEST_CASE("Unit_hipFreeDoubleDevice") {
+  size_t width = GENERATE(32, 512, 1024);
+  char* ptr{};
+  size_t size_mult = width;
+  HIP_CHECK(hipMalloc(&ptr, sizeof(char) * size_mult));
+
+  HIP_CHECK(hipFree(ptr));
+  HIP_CHECK_ERROR(hipFree(ptr), hipErrorInvalidValue);
+}
+TEST_CASE("Unit_hipFreeDoubleHost") {
+  size_t width = GENERATE(32, 512, 1024);
+  char* ptr{};
+  size_t size_mult = width;
+
+  HIP_CHECK(hipHostMalloc(&ptr, sizeof(char) * size_mult));
+
+  HIP_CHECK(hipHostFree(ptr));
+  HIP_CHECK_ERROR(hipHostFree(ptr), hipErrorInvalidValue);
+}
+TEST_CASE("Unit_hipFreeDoubleArrayFree") {
+#if HT_NVIDIA
+  HipTest::HIP_SKIP_TEST("EXSWCPHIPT-120");
+  return;
+#endif
+  size_t width = GENERATE(32, 512, 1024);
+  size_t height = GENERATE(0, 32, 512, 1024);
+  hipArray_t arrayPtr{};
+  hipExtent extent{};
+  extent.width = width;
+  extent.height = height;
+  hipChannelFormatDesc desc = hipCreateChannelDesc<char>();
+
+  HIP_CHECK(hipMallocArray(&arrayPtr, &desc, extent.width, extent.height, hipArrayDefault));
+
+  HIP_CHECK(hipFreeArray(arrayPtr));
+  HIP_CHECK_ERROR(hipFreeArray(arrayPtr), hipErrorContextIsDestroyed);
+}
+#if HT_NVIDIA
+TEST_CASE("Unit_hipFreeDoubleArrayDestroy") {
+#if HT_NVIDIA
+  HipTest::HIP_SKIP_TEST("EXSWCPHIPT-120");
+  return;
+#endif
+  using vec_info = vector_info<char>;
+
+  size_t width = GENERATE(32, 512, 1024);
+  size_t height = GENERATE(0, 32, 512, 1024);
+  DriverContext ctx{};
+
+  hiparray ArrayPtr{};
+  HIP_ARRAY_DESCRIPTOR cuDesc;
+  cuDesc.Width = width;
+  cuDesc.Height = height;
+  cuDesc.Format = vec_info::format;
+  cuDesc.NumChannels = vec_info::size;
+  HIP_CHECK(hipArrayCreate(&ArrayPtr, &cuDesc));
+  HIP_CHECK(hipArrayDestroy(ArrayPtr));
+  HIP_CHECK_ERROR(hipArrayDestroy(ArrayPtr), hipErrorContextIsDestroyed);
+}
+#endif
+
+
+// DevFree, ArrayFree, ArrayDestroy, HostFree Mutithreaded
+TEMPLATE_TEST_CASE("Unit_hipFreeMultiTDev", "", char, int, float2, float4) {
+  constexpr size_t numAllocs = 10;
+  std::vector<TestType*> ptrs(numAllocs);
+  size_t allocSize = sizeof(TestType) * GENERATE(1, 32, 64, 128);
+
+  for (auto& ptr : ptrs) {
+    HIP_CHECK(hipMalloc(&ptr, allocSize));
+  }
+
+  std::vector<std::thread> threads;
+
+  for (auto ptr : ptrs) {
+    threads.emplace_back(std::thread([ptr] {
```

Comment:
```suggestion
    threads.emplace_back([ptr] {
```
emplace_back should forward the argument to the constructor of the type.

Reply:
Amended